Sprint-2/extend/format-time.js

```diff
@@ -1,24 +1,94 @@
 // This is the latest solution to the problem from the prep.
+// https://programming.codeyourfuture.io/structuring-data/sprints/2/prep/
 // Your task is to write tests for as many different groups of input data or edge cases as you can, and fix any bugs you find.
 
 function formatAs12HourClock(time) {
+  // check my time is valid first
+  validateTimeInput(time);
   const hours = Number(time.slice(0, 2));
+  // handle midnight
+  if (hours === 0) return "12:00 am";
+  // handle noon
+  if (hours === 12) return "12:00 pm";
+  // otherwise convert to 12-hour clock
   if (hours > 12) {
     return `${hours - 12}:00 pm`;
   }
   return `${time} am`;
 }
 
-const currentOutput = formatAs12HourClock("08:00");
-const targetOutput = "08:00 am";
-console.assert(
-  currentOutput === targetOutput,
-  `current output: ${currentOutput}, target output: ${targetOutput}`
-);
-
-const currentOutput2 = formatAs12HourClock("23:00");
-const targetOutput2 = "11:00 pm";
-console.assert(
-  currentOutput2 === targetOutput2,
-  `current output: ${currentOutput2}, target output: ${targetOutput2}`
-);
+function validateTimeInput(time) {
+  // 1. A time string should be exactly 5 characters like "14:30"
+  if (time.length !== 5) {
+    throw new Error("Time must be in format HH:MM");
+  }
+
+  // 2. Parse hours and minutes - this uses destructuring
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
+  const [hours, minutes] = time.split(":").map(Number);
+
+  // 3. Check if we got valid numbers within range
+  if (
+    hours === undefined ||
+    minutes === undefined ||
+    hours > 23 ||
+    minutes > 59
+  ) {
+    throw new Error("Invalid time - use hours 00-23 and minutes 00-59");
```

Comment:
Your error message suggests there are lower bounds on these values, but your if statement doesn't check for them :)

Reply:
What are you expecting to change next sprint? I don't think there's a "revise your answers" thing in sprint 3? (Or honestly that there's more relevant knowledge next sprint - sprint 3 really just teaches Jest...)

(Also, FWIW, I did not originally expect validation / throwing to be a part of this exercise, but I'm not unhappy with it! But I do think adding a `hours < 0 || minutes < 0` check to the `if` is reasonable with the knowledge trainees have at this point)